lib/Sema/CSSyntacticElement.cpp

```diff
@@ -211,6 +211,14 @@ class TypeVariableRefFinder : public ASTWalker {
         return;
     }
 
+    // Desugar type before collecting type variables, otherwise
+    // we can bring in scope unrelated type variables passed
+    // into the closure (via parameter/result) from contextual type.
+    // For example `Typealias<$T, $U>.Context` which desugars into
+    // `_Context<$U>` would bring in `$T` that could be inferrable
+    // only after the body of the closure is solved.
+    type = type->getDesugaredType();
```

Comment:
I think getDesugaredType() only unwraps one level of sugar, but you might have type aliases in nested position. Maybe you should use getCanonicalType() instead? Are other callers of getTypeVariables() also affected by this? Perhaps getTypeVariables() should canonicalize the type itself first.

Reply:
Indeed, getDesugaredType would only unwrap N levels of sugar if they occur in sequence... I'll switch this to `getCanonicalType()`.